c++/src/capnp/compiler/capnpc-c++.c++

```diff
@@ -2122,10 +2122,22 @@ private:
     auto interfaceIdHex = kj::hex(interfaceId);
     uint16_t methodId = method.getIndex();
 
+    // TODO(msvc):  Notice that the return type of this method's request function is supposed to be
+    // `::capnp::Request<param, result>`. If the first template parameter to ::capnp::Request is a
+    // template instantiation, MSVC will sometimes complain that it's unspecialized and can't be
+    // used as a parameter in the return type (error C3203). It is not clear to me under what exact
+    // conditions this bug occurs, but it commonly crops up in test.capnp.h.
+    //
+    // The easiest (and only) workaround I found is to use C++14's return type deduction here, thus
+    // the `#if _MSC_VER` hackery in the return type declarations below. We're depending on the fact
+    // that that this function has an inline implementation for the deduction to work.
+
     auto requestMethodImpl = kj::strTree(
         templateContext.allDecls(),
         implicitParamsTemplateDecl,
-        "::capnp::Request<", paramType, ", ", resultType, ">\n",
+        templateContext.isGeneric() ? "#if _MSC_VER\nauto\n#else\n" : "",
```

Comment:
Instead of littering the generated code with these verbose if/elses, what about defining a macro like `CAPNP_AUTO_IF_MSVC(type)` which evaluates to `auto` on MSVC or `type` elsewhere?